### PR TITLE
test: adding synchronization for tap tests

### PR DESCRIPTION
**Pull request by alyssawilk**
Hopefully fixes https://github.com/envoyproxy/envoy/issues/26258 (see issue for failure details)

### CODE REVIEW

File: source/extensions/common/tap/extension_config_base.cc [hunk 1]
```diff
@@ -65,9 +65,11 @@ void ExtensionConfigBase::installNewTap(const envoy::config::tap::v3::TapConfig&
...
   TapConfigSharedPtr new_config =
       config_factory_->createConfigFromProto(proto_config, admin_streamer);
-  tls_slot_.runOnAllThreads([new_config](OptRef<TlsFilterConfig> tls_filter_config) {
-    tls_filter_config->config_ = new_config;
-  });
+  tls_slot_.runOnAllThreads(
+      [new_config](OptRef<TlsFilterConfig> tls_filter_config) {
+        tls_filter_config->config_ = new_config;
+      },
+      []() -> void { ENVOY_LOG(debug, "New tap installed on all workers."); });
```
Comment (RyanTheOptimist) [inline source/extensions/common/tap/extension_config_base.cc:72]: Out of curiosity, would it work to move the ENVOY_LOG() into the body of the first lambda, after line 70?

File: test/test_common/logging.h [hunk 2]
```diff
@@ -201,4 +201,31 @@ using ExpectedLogMessages = std::vector<StringPair>;
     ASSERT_EQ(0, logs.size()) << " Logs:\n   " << absl::StrJoin(logs, "   ");                      \
   } while (false)
 
+// Validates that when stmt is executed, the supplied substring is eventually logged.
+// This both waits infinitely (not ideal) and walks the full log list every time (not ideal)
```
Comment (RyanTheOptimist) [inline test/test_common/logging.h:205]: But other than that, this is ideal, right? :P In any case, nit: please add a "." to the end of the comment.

**Review by RyanTheOptimist — APPROVED** [review]

File: source/extensions/common/tap/extension_config_base.cc (hunk 1 above)
Comment (alyssawilk) [reply source/extensions/common/tap/extension_config_base.cc:72]: nope.  runonAllThreads is async, and allows for a callOnComplete closure, which happens  after all the posted closures are actually run 
If we moved it out we'd have the same problem where we'd log after we called runOnAllThreads but potentially before those async posted callbacks had actually run on individual worker threads

Comment (RyanTheOptimist) [reply source/extensions/common/tap/extension_config_base.cc:72]: Oh! I see. Run on *all* threads. Gotcha.

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by alyssawilk** [comment]
when your flake fix fails CI due to different flakes  :-(
/retest

**Pull request merged**